R/Plot_synergy.R

```diff
@@ -1,328 +1,330 @@
-      
-      if(!is.null(col.range)) {
-        xconc <- round(col.conc[col.range[1]:col.range[2]], 3)
-      } else {
-        xconc <- round(col.conc, 3)
-      }
-      
-      axis(side = 1, at = seq(0, 1, by = 1/(length(xconc) - 1)), labels = xconc)
-      axis(side = 2, at = seq(0, 1, by = 1/(length(yconc) - 1)), labels = yconc)
-      print(syn.3d.plot, position = c(0.5,0, 1, 1), newpage = FALSE)
-      fig <- recordPlot()
-    }
-    plots[[i]] <- fig
-    if(save.file) {
-      if (type %in% c("2D", "3D")) {
-        pdf(file.name, width = 10, height = 10)
-      } else {
-        pdf(file.name, width = 12, height = 6)
-      }
-      print(fig)
-      dev.off()
-    }
+# Returns nested associative lists with all data needed for plots of any type.
+.GetPlotData <- function(
+  data, index, legend.start, legend.end, row.range, col.range, len) {
+  scores.dose <- data$scores[[index]]
+  color.range <- .RoundUpToNextTen(scores.dose)
+  unit.text <- paste("(", data$drug.pairs$concUnit[index], ")", sep = "")
+  return(list(
+    metadata = list(  # useful for naming output files
+      drug.blockID = data$drug.pairs$blockIDs[index],
+      file.name.prefix = paste(data$drug.pairs$drug.row[index],
+                               data$drug.pairs$drug.col[index], sep = '.')
+    ),
+    data = list(  # useful for actual plot content
+      plot.title =
+        .ComputePlotTitle(scores.dose, row.range, col.range, data$method),
+      # Note, when considering application of ranges, that with len=3 mat.tmp will
+      # have 9 times as many elements as scores.dose.  It is best to invoke
+      # .ExtendedScores only after ranges have been applied.
+      mat.tmp =
+        .ExtendedScores(.ApplyRanges(scores.dose, row.range, col.range), len),
+      len = len,
+      row.conc = .NumericApplyRange(rownames(scores.dose), row.range),
+      col.conc = .NumericApplyRange(colnames(scores.dose), col.range),
+      drug.row = paste(data$drug.pairs$drug.row[index], unit.text, sep = " "),
+      drug.col = paste(data$drug.pairs$drug.col[index], unit.text, sep = " "),
+      start.point = if(is.null(legend.start)) -color.range else legend.start,
+      end.point = if(is.null(legend.end)) color.range else legend.end)))
+}
 
-  }
+# Returns the highest absolute value in 'scores', rounded up to next ten.
+.RoundUpToNextTen <- function(scores) {
+  return(round(5 + max(abs(max(scores)),
+                       abs(min(scores))),
+               -1))
+}
+
+# Returns a string that may be used as a plot title.  The string includes
+# the mean value of all scores in the selected ranges, and the name of the
+# method used to compute them.
+.ComputePlotTitle <- function(scores.dose, row.range, col.range, data.method) {
+  # The first row and column are removed no matter what,
+  # presumably (? TODO/REVIEW) on the assumptions that:
+  #  1. For each drug, the first recorded effect is for dose=0.
+  #  2. When dose=0 for either drug, the synergy score is also 0.
```

Comment:
when I was writing the code, I wanted to keep the score matrix and dose-response matrix the same size. So I filled the first row and column with zeros. But when I think about it now, it seems not necessary to keep them in the same size.

when dose = 0 for either drug (the first entry of the matrix) meaning no drugs in the well,  negative control DMSO is in the well. DMSO is the solution for drugs.

Reply:
Acknowledged (see response just above).